src/basic/process-util.c

```diff
@@ -158,51 +165,103 @@ int get_process_cmdline(pid_t pid, size_t max_columns, ProcessCmdlineFlags flags
+                                if (t[i] == '\0')
+                                        t[i] = ' ';
 
-                if (!(flags & PROCESS_CMDLINE_COMM_FALLBACK))
-                        return -ENOENT;
+                        t[k] = '\0'; /* Normally, t[k] is already NUL, so this is just a guard in case of short read */
 
-                /* Kernel threads have no argv[] */
-                _cleanup_free_ char *t2 = NULL;
+                        delete_trailing_chars(t, WHITESPACE);
 
-                r = get_process_comm(pid, &t2);
-                if (r < 0)
-                        return r;
+                } else {
+                        /* We only treat getting nothing as an error. We *could* also get an error after
+                         * reading some data, but we ignore that case, as such an error is rather unlikely
+                         * and we prefer to get some data rather than none. */
+                        if (ferror(f))
+                                return -errno;
 
-                mfree(t);
-                t = strjoin("[", t2, "]");
-                if (!t)
+                        if (!(flags & PROCESS_CMDLINE_COMM_FALLBACK))
+                                return -ENOENT;
+
+                        /* Kernel threads have no argv[] */
+                        _cleanup_free_ char *t2 = NULL;
+
+                        r = get_process_comm(pid, &t2);
+                        if (r < 0)
+                                return r;
+
+                        mfree(t);
+                        t = strjoin("[", t2, "]");
+                        if (!t)
+                                return -ENOMEM;
+                }
+
+                bool eight_bit = (flags & PROCESS_CMDLINE_USE_LOCALE) && !is_locale_utf8();
+
+                ans = escape_non_printable_full(t, max_columns, eight_bit);
+                if (!ans)
                         return -ENOMEM;
-        }
 
-        delete_trailing_chars(t, WHITESPACE);
+                ans = str_realloc(ans);
 
-        bool eight_bit = (flags & PROCESS_CMDLINE_USE_LOCALE) && !is_locale_utf8();
+        } else {
+                _cleanup_strv_free_ char **args = NULL;
 
-        ans = escape_non_printable_full(t, max_columns, eight_bit);
-        if (!ans)
-                return -ENOMEM;
+                assert(!(flags & PROCESS_CMDLINE_USE_LOCALE));
+
+                for (;;) {
+                        _cleanup_free_ char *arg = NULL, *escaped = NULL;
+
+                        r = read_nul_string(f, max_length, &arg);
+                        if (r < 0)
+                                return r;
+                        if (r == 0)
+                                break;
+
+                        escaped = shell_maybe_quote(arg, SHELL_ESCAPE_EMPTY);
+                        if (!escaped)
+                                return -ENOMEM;
+
+                        r = strv_consume(&args, TAKE_PTR(escaped) ?: TAKE_PTR(arg));
```

Comment:
you checked that escaped is non-NULL the line before, so the ternary op doesn't make much sense?